File: sdks/typescript/src/stream.ts
import type { Event, Response as PalaiResponse } from "./generated/types.ts";
import { errorForResponse, PalaiConnectionError } from "./errors.ts";

// A parsed Server-Sent Events frame (WHATWG event stream format). `data` is the joined
// data lines; `id` updates the reconnection cursor; `event` is the event name the server
// set. A frame with neither data nor event (a bare heartbeat comment) is never emitted.
export interface SSEFrame {
  id: string | undefined;
  event: string | undefined;
  data: string;
}

// parseEventStream frames a byte stream into SSE frames. It decodes incrementally
// (TextDecoder stream mode preserves a multi-byte character split across chunks), buffers
// a partial trailing line across reads, joins multi-line `data:` fields with newlines, and
// ignores comment lines (a leading colon), which the server uses for keep-alive. It never
// buffers the whole response — one frame is produced per blank-line dispatch.
export async function* parseEventStream(
  body: ReadableStream<Uint8Array>,
): AsyncGenerator<SSEFrame> {
  const reader = body.getReader();
  const decoder = new TextDecoder();
  let buffer = "";
  let frame = newFrame();
  try {
    for (;;) {
      const { done, value } = await reader.read();
      if (done) {
        break;
      }
      buffer += decoder.decode(value, { stream: true });
      let newline: number;
      while ((newline = buffer.indexOf("\n")) !== -1) {
        // Tolerate CRLF as well as LF line endings.
        let line = buffer.slice(0, newline);
        if (line.endsWith("\r")) {
          line = line.slice(0, -1);
        }
        buffer = buffer.slice(newline + 1);
        if (line === "") {
          if (frame.data !== "" || frame.event !== undefined || frame.id !== undefined) {
            yield frame;
          }
          frame = newFrame();
          continue;
        }
        applyField(frame, line);
      }
    }
  } finally {
    // Releasing the reader lets the caller cancel the body to close the socket.
    reader.releaseLock();
  }
}

interface MutableFrame {
  id: string | undefined;
  event: string | undefined;
  data: string;
}

function newFrame(): MutableFrame {
  return { id: undefined, event: undefined, data: "" };
}

// applyField parses one SSE line into the current frame. A leading colon is a comment; a
// field is `name` or `name:value` with one optional space stripped after the colon.
function applyField(frame: MutableFrame, line: string): void {
  if (line.startsWith(":")) {
    return; // comment / heartbeat
  }
  const colon = line.indexOf(":");
  const field = colon === -1 ? line : line.slice(0, colon);
  let value = colon === -1 ? "" : line.slice(colon + 1);
  if (value.startsWith(" ")) {
    value = value.slice(1);
  }
  switch (field) {
    case "id":
      frame.id = value;
      break;
    case "event":
      frame.event = value;
      break;
    case "data":
      frame.data = frame.data === "" ? value : `${frame.data}\n${value}`;
      break;
    default:
      // retry: and unknown fields are not used by this consumer.
      break;
  }
}

const TERMINAL_EVENT = /^(run|response)\.(completed|failed|canceled|timed_out|budget_exceeded)\.v[0-9]+$/;

// isTerminalEvent reports whether an event closes the run, so the stream stops rather than
// reconnecting when the server closes a completed stream (spec §22.3, §24.4).
export function isTerminalEvent(event: Event): boolean {
  return TERMINAL_EVENT.test(event.type);
}

// StreamTransport is the slice of the client the stream needs: it opens the raw SSE
// response (Bearer auth + Last-Event-ID applied by the client) and retrieves the canonical
// terminal Response. Kept structural so stream.ts does not import the client (no cycle).
export interface StreamTransport {
  openEventStream(
    sessionID: string,
    lastEventId: string | null,
    signal: AbortSignal | undefined,
  ): Promise<globalThis.Response>;
  retrieveResponse(responseID: string): Promise<PalaiResponse>;
}

// StreamStart is the identity a stream resolves to when its underlying response is
// created (or already exists): the response to retrieve at the end and the session to
// subscribe to.
export interface StreamStart {
  responseID: string;
  sessionID: string;
}

export interface ResponseStreamInit {
  transport: StreamTransport;
  /** Runs once, lazily, on first consumption — typically the create POST. */
  start: () => Promise<StreamStart>;
  signal?: AbortSignal | undefined;
  lastEventId?: string | null;
  maxReconnects?: number;
  backoffBaseMs?: number;
  backoffMaxMs?: number;
}

// ResponseStream is a resumable, typed consumer of a run's event stream. Iterating it
// yields each canonical Event (unknown event types are delivered, not dropped: API-009).
// A transport drop before a terminal event reconnects from the last seen id via
// Last-Event-ID with full-jitter backoff; a terminal event or an explicit cancel stops
// without reconnecting. Breaking out of the iteration closes the transport.
export class ResponseStream implements AsyncIterable<Event> {
  #transport: StreamTransport;
  #start: () => Promise<StreamStart>;
  #started: StreamStart | undefined;
  #startPromise: Promise<StreamStart> | undefined;
  #signal: AbortSignal | undefined;
  #lastEventId: string | null;
  #maxReconnects: number;
  #backoffBaseMs: number;
  #backoffMaxMs: number;
  #iterator: AsyncGenerator<Event> | undefined;

  constructor(init: ResponseStreamInit) {
    this.#transport = init.transport;
    this.#start = init.start;
    this.#signal = init.signal;
    this.#lastEventId = init.lastEventId ?? null;
    this.#maxReconnects = init.maxReconnects ?? 5;
    this.#backoffBaseMs = init.backoffBaseMs ?? 100;
    this.#backoffMaxMs = init.backoffMaxMs ?? 5_000;
  }

  // responseID / sessionID resolve once the stream has started (its create has returned);
  // they are undefined before first consumption.
  get responseID(): string | undefined {
    return this.#started?.responseID;
  }
  get sessionID(): string | undefined {
    return this.#started?.sessionID;
  }

  // lastEventId exposes the reconnection cursor reached so far (the id of the last event
  // delivered), so a caller can persist it and resume a new stream later.
  get lastEventId(): string | null {
    return this.#lastEventId;
  }

  #ensureStarted(): Promise<StreamStart> {
    if (this.#startPromise === undefined) {
      this.#startPromise = this.#start().then((started) => (this.#started = started));
    }
    return this.#startPromise;
  }

  [Symbol.asyncIterator](): AsyncGenerator<Event> {
    if (this.#iterator !== undefined) {
      throw new PalaiConnectionError("this response stream has already been consumed");
    }
    this.#iterator = this.#run();
    return this.#iterator;
  }

  // finalResponse drains the stream to its terminal event, then returns the canonical
  // terminal Response (status, model, output, usage, and — on failure — the problem-shaped
  // error). It shares the single underlying iteration, so it must not be combined with a
  // separate manual iteration of the same stream.
  async finalResponse(): Promise<PalaiResponse> {
    const iterator = this[Symbol.asyncIterator]();
    for (;;) {
      const next = await iterator.next();
      if (next.done) {
        break;
      }
      if (isTerminalEvent(next.value)) {
        break;
      }
    }
    await iterator.return?.(undefined);
    const { responseID } = await this.#ensureStarted();
    return this.#transport.retrieveResponse(responseID);
  }

  async *#run(): AsyncGenerator<Event> {
    let start: StreamStart;
    try {
      start = await this.#ensureStarted();
    } catch (cause) {
      if (isAbort(cause, this.#signal)) {
        return; // the create was canceled before the stream opened
      }
      throw cause; // a create failure surfaces to the consumer
    }
    let reconnects = 0;
    for (;;) {
      // Capture the cursor this connection resumes from, so a server that redelivers the
      // boundary event inclusively does not double-deliver it to the consumer.
      const resumedFrom = this.#lastEventId;
      let response: globalThis.Response;
      try {
        response = await this.#transport.openEventStream(start.sessionID, resumedFrom, this.#signal);
      } catch (cause) {
        if (isAbort(cause, this.#signal)) {
          return; // explicit cancel: stop, do not reconnect
        }
        if (reconnects >= this.#maxReconnects) {
          throw new PalaiConnectionError("event stream could not be (re)opened", { cause });
        }
        if (!(await this.#backoffSleep(reconnects))) {
          return; // canceled mid-backoff: end quietly, like every other cancel path
        }
        reconnects += 1;
        continue;
      }
      if (!response.ok) {
        // A status error (404 unknown session, 410 gone) is terminal, not a drop.
        const body = await safeText(response);
        throw errorForResponse(response.status, body, response.headers.get("Request-Id") ?? undefined);
      }
      if (response.body === null) {
        return;
      }
      const frames = parseEventStream(response.body);
      // On a resume, drop a single redelivered boundary event whose id equals the cursor.
      let dedupePending = resumedFrom !== null;
      try {
        for await (const frame of frames) {
          if (frame.id !== undefined) {
            this.#lastEventId = frame.id;
          }
          if (frame.data === "") {
            continue; // keep-alive with no payload
          }
          const event = decodeEvent(frame.data);
          if (event === null) {
            continue; // a non-JSON data line is not a canonical event
          }
          if (dedupePending) {
            dedupePending = false;
            if (frame.id !== undefined && frame.id === resumedFrom) {
              continue; // duplicate of the resume boundary
            }
          }
          yield event;
          if (isTerminalEvent(event)) {
            return; // terminal reached: stop, do not reconnect
          }
        }
      } catch (cause) {
        if (isAbort(cause, this.#signal)) {
          return;
        }
        // fall through to reconnect below
      } finally {
        // Close the body — and thus the socket — whenever this connection is left: on a
        // terminal, an abort, or the consumer breaking out of the iteration (iterator
        // close closes the transport). It is a no-op when the server already ended it.
        await frames.return(undefined);
        await response.body?.cancel().catch(() => {});
      }
      if (this.#signal?.aborted) {
        return;
      }
      // The stream ended without a terminal event: an unexpected drop. Reconnect from the
      // last seen id, bounded, with backoff.
      if (reconnects >= this.#maxReconnects) {
        throw new PalaiConnectionError("event stream dropped before a terminal event and exhausted reconnects");
      }
      if (!(await this.#backoffSleep(reconnects))) {
        return; // canceled mid-backoff: end quietly, like every other cancel path
      }
      reconnects += 1;
    }
  }

  // #backoffSleep waits out one reconnect backoff, cancelably. It returns true when the
  // sleep completes, or false if the caller aborts mid-sleep — so #run ends the stream
  // quietly instead of leaking the abort's DOMException, matching every other cancel path.
  async #backoffSleep(reconnects: number): Promise<boolean> {
    try {
      await delay(fullJitterBackoff(reconnects, this.#backoffBaseMs, this.#backoffMaxMs), this.#signal);
      return true;
    } catch (cause) {
      if (isAbort(cause, this.#signal)) {
        return false;
      }
      throw cause;
    }
  }
}

function decodeEvent(data: string): Event | null {
  try {
    const parsed = JSON.parse(data) as unknown;
    if (typeof parsed === "object" && parsed !== null && typeof (parsed as Event).type === "string") {
      return parsed as Event;
    }
    return null;
  } catch {
    return null;
  }
}

async function safeText(response: globalThis.Response): Promise<string> {
  try {
    return await response.text();
  } catch {
    return "";
  }
}

// isAbort reports whether an error is the caller's explicit cancellation, so the stream
// stops silently rather than surfacing an error or reconnecting.
function isAbort(cause: unknown, signal: AbortSignal | undefined): boolean {
  if (signal?.aborted) {
    return true;
  }
  return cause instanceof Error && cause.name === "AbortError";
}

// --- Shared backoff, used by both stream reconnect and the client's request retry. ---

// fullJitterBackoff returns a delay in [0, min(maxMs, baseMs * 2^attempt)] — the AWS
// "full jitter" schedule, which spreads retries so many clients do not synchronize their
// reconnect storms (§23.7). attempt is 0-based.
export function fullJitterBackoff(attempt: number, baseMs: number, maxMs: number): number {
  if (baseMs <= 0 || attempt < 0) {
    return 0;
  }
  const ceiling = Math.min(maxMs, baseMs * 2 ** attempt);
  return Math.random() * ceiling;
}

// delay resolves after ms, or rejects with the signal's reason if it aborts first, so a
// backoff sleep is itself cancelable.
export function delay(ms: number, signal?: AbortSignal | undefined): Promise<void> {
  return new Promise((resolve, reject) => {
    if (signal?.aborted) {
      reject(signal.reason);
      return;
    }
    const timer = setTimeout(resolve, ms);
    signal?.addEventListener(
      "abort",
      () => {
        clearTimeout(timer);
        reject(signal.reason);
      },
      { once: true },
    );
  });
}

File: sdks/go/stream.go
package palai

import (
	"bufio"
	"context"
	"encoding/json"
	"io"
	"math/rand"
	"net/http"
	"regexp"
	"strings"
	"time"
)

// The streaming layer: a WHATWG SSE frame parser mirroring the TS SDK's parseEventStream (and the
// conformance harness's reference framer) exactly, plus ResponseStream — a resumable, typed
// consumer that reconnects from Last-Event-ID on a drop and stops on a terminal event.

// maxSSELineBytes caps a single SSE line, mirroring provider_one's 1 MiB frame ceiling SDK-side so
// a hostile or runaway stream cannot exhaust memory on one unbounded line.
const maxSSELineBytes = 1 << 20

// SSEFrame is a parsed Server-Sent Events frame. Data is the joined data lines; ID updates the
// reconnection cursor; Event is the event name the server set.
type SSEFrame struct {
	ID    string
	Event string
	Data  string
}

// scanSSE frames an SSE byte stream, invoking fn for each dispatched frame (blank-line boundary),
// stopping early if fn returns false. It tolerates CRLF, skips comment lines (a leading colon),
// joins multi-line data fields with "\n", strips one optional space after a field colon, and caps
// a single line at maxSSELineBytes. It returns the first read error (nil at clean EOF).
//
// A trailing frame not terminated by a blank line is DISCARDED at EOF, mirroring parseEventStream,
// the harness reference framer, and WHATWG: a graceful close mid-frame (e.g. a Caddy-edge idle FIN)
// must not dispatch a truncated event, or a resume would advance Last-Event-ID past an event that
// was never delivered.
func scanSSE(r io.Reader, fn func(SSEFrame) bool) error {
	sc := bufio.NewScanner(r)
	// bufio.ScanLines splits on '\n' and strips a trailing '\r', so CRLF is tolerated. The buffer
	// cap is the 1 MiB frame ceiling: a line past it surfaces as bufio.ErrTooLong, not OOM.
	sc.Buffer(make([]byte, 0, 64*1024), maxSSELineBytes)
	var (
		frame   SSEFrame
		hasData bool
		hasAny  bool
	)
	for sc.Scan() {
		line := sc.Text()
		if line == "" { // blank line dispatches the buffered frame
			if hasAny && !fn(frame) {
				return nil
			}
			frame, hasData, hasAny = SSEFrame{}, false, false
			continue
		}
		hasAny = true
		applySSEField(&frame, &hasData, line)
	}
	return sc.Err()
}

// applySSEField parses one SSE line into the current frame. A leading colon is a comment; a field
// is `name` or `name:value` with one optional space stripped after the colon.
func applySSEField(frame *SSEFrame, hasData *bool, line string) {
	if strings.HasPrefix(line, ":") {
		return // comment / heartbeat
	}
	field, value := line, ""
	if colon := strings.IndexByte(line, ':'); colon != -1 {
		field, value = line[:colon], strings.TrimPrefix(line[colon+1:], " ")
	}
	switch field {
	case "id":
		frame.ID = value
	case "event":
		frame.Event = value
	case "data":
		if *hasData {
			frame.Data += "\n" + value
		} else {
			frame.Data = value
			*hasData = true
		}
	default:
		// retry: and unknown fields are not used by this consumer.
	}
}

var terminalEventRe = regexp.MustCompile(`^(run|response)\.(completed|failed|canceled|timed_out|budget_exceeded)\.v[0-9]+$`)

// IsTerminalEvent reports whether an event closes the run, so the stream stops rather than
// reconnecting (spec §22.3, §24.4).
func IsTerminalEvent(e Event) bool {
	return terminalEventRe.MatchString(e.Type)
}

// ScanEvents frames an SSE byte stream and delivers each decoded canonical Event to fn (unknown
// event types and fields preserved; comments, heartbeats, and non-JSON data lines skipped), stopping
// early if fn returns false. It is the low-level parser ResponseStream builds on, exported for a
// caller that holds a raw event-stream body directly.
func ScanEvents(r io.Reader, fn func(Event) bool) error {
	return scanSSE(r, func(f SSEFrame) bool {
		if f.Data == "" {
			return true
		}
		e, ok := decodeEvent(f.Data)
		if !ok {
			return true
		}
		return fn(e)
	})
}

// decodeEvent decodes an SSE data line into a canonical Event iff it is a JSON object with a
// string `type`; otherwise it returns ok=false (a non-JSON heartbeat or plain line).
func decodeEvent(data string) (Event, bool) {
	var probe struct {
		Type *string `json:"type"`
	}
	if json.Unmarshal([]byte(data), &probe) != nil || probe.Type == nil {
		return Event{}, false
	}
	var e Event
	if json.Unmarshal([]byte(data), &e) != nil {
		return Event{}, false
	}
	return e, true
}

// fullJitterBackoff returns a delay in [0, min(maxMs, baseMs*2^attempt)] — the AWS "full jitter"
// schedule, which spreads retries so many clients do not synchronize their reconnect storms
// (spec §23.7). attempt is 0-based.
func fullJitterBackoff(attempt, baseMs, maxMs int) time.Duration {
	if baseMs <= 0 || attempt < 0 {
		return 0
	}
	ceiling := maxMs
	if exp := baseMs << attempt; attempt < 31 && exp < ceiling {
		ceiling = exp
	}
	return time.Duration(rand.Intn(ceiling+1)) * time.Millisecond
}

// sleep waits d or returns ctx.Err() if the context ends first, so a backoff is cancelable.
func sleep(ctx context.Context, d time.Duration) error {
	if d <= 0 {
		return ctx.Err()
	}
	t := time.NewTimer(d)
	defer t.Stop()
	select {
	case <-ctx.Done():
		return ctx.Err()
	case <-t.C:
		return nil
	}
}

// streamTransport is the seam a shared reconnect/backoff/terminal engine (runStream) is built on.
// Only the connect strategy and per-frame cursor bookkeeping differ between ResponseStream
// (Last-Event-ID header + a dedupe check on the resume boundary, see responseStreamTransport in this
// file) and SessionEventStream (?after_sequence, no dedupe needed — the server's read is
// strictly-greater-than, see sessionStreamTransport in sessionevents.go). Everything else — backoff,
// terminal handling, non-2xx and exhausted-reconnect errors — is identical and lives once in
// runStream.
type streamTransport interface {
	// open connects (or reconnects), resuming from whatever cursor state the transport holds.
	open(ctx context.Context) (*http.Response, error)
	// processFrame turns one raw SSE frame into a decoded Event ready to deliver (ok=false for a
	// non-JSON/heartbeat frame or a dedup boundary), updating the transport's own resume cursor as a
	// side effect. It is called for EVERY dispatched frame, decoded or not, so an id-only frame still
	// advances an id-keyed cursor.
	processFrame(f SSEFrame) (Event, bool)
}

// runStream is the reconnect/backoff/terminal engine shared by ResponseStream.Events and
// SessionEventStream.run: open a connection (or reuse first, an already-open one a caller needing a
// synchronous connect error opened itself — nil for a caller with no such requirement), scan it with
// scanSSE, deliver each decoded event, stop cleanly on a terminal event or a deliver-requested stop,
// and otherwise reconnect with full-jitter backoff (bounded by maxReconnects) — on an open failure, a
// non-2xx open (immediate, no retry), or the connection simply ending (a clean EOF or a read error;
// scanSSE's returned error is not distinguished between the two, matching both streams' original
// behavior pre-extraction).
func runStream(ctx context.Context, t streamTransport, first *http.Response, maxReconnects, backoffBaseMs, backoffMaxMs int, deliver func(Event, error) bool) {
	reconnects := 0
	resp := first
	for {
		if resp == nil {
			opened, err := t.open(ctx)
			if err != nil {
				if ctx.Err() != nil {
					return
				}
				if reconnects >= maxReconnects {
					deliver(Event{}, &ConnectionError{Message: "event stream could not be (re)opened", Cause: err})
					return
				}
				if sleep(ctx, fullJitterBackoff(reconnects, backoffBaseMs, backoffMaxMs)) != nil {
					return
				}
				reconnects++
				continue
			}
			if opened.StatusCode/100 != 2 {
				body, _ := io.ReadAll(io.LimitReader(opened.Body, maxSSELineBytes))
				opened.Body.Close()
				deliver(Event{}, errorForResponse(opened.StatusCode, string(body), opened.Header.Get("Request-Id")))
				return
			}
			resp = opened
		}

		terminal, stopped := false, false
		_ = scanSSE(resp.Body, func(f SSEFrame) bool {
			e, ok := t.processFrame(f)
			if !ok {
				return true
			}
			if !deliver(e, nil) {
				stopped = true
				return false
			}
			if IsTerminalEvent(e) {
				terminal = true
				return false
			}
			return true
		})
		resp.Body.Close()
		resp = nil // this connection is spent either way; a next iteration (re)opens
		if terminal || stopped || ctx.Err() != nil {
			return
		}
		if reconnects >= maxReconnects {
			deliver(Event{}, &ConnectionError{Message: "event stream dropped before a terminal event and exhausted reconnects"})
			return
		}
		if sleep(ctx, fullJitterBackoff(reconnects, backoffBaseMs, backoffMaxMs)) != nil {
			return
		}
		reconnects++
	}
}

// ResponseStream is a resumable, typed consumer of a run's event stream. Iterating it via Events
// yields each canonical Event (unknown event types are delivered, not dropped: API-009). A
// transport drop before a terminal event reconnects from the last seen id via Last-Event-ID with
// full-jitter backoff; a terminal event stops without reconnecting.
type ResponseStream struct {
	client        *Client
	responseID    string
	sessionID     string
	lastEventID   string
	maxReconnects int
	backoffBaseMs int
	backoffMaxMs  int
}

// LastEventID exposes the reconnection cursor reached so far, so a caller can persist and resume.
func (s *ResponseStream) LastEventID() string { return s.lastEventID }

// ResponseID / SessionID identify the run this stream follows.
func (s *ResponseStream) ResponseID() string { return s.responseID }
func (s *ResponseStream) SessionID() string  { return s.sessionID }

// responseStreamTransport adapts ResponseStream to streamTransport: resume is the Last-Event-ID
// header (an evt_* id). A resumed connection's first decoded event is suppressed if its frame id
// equals the resume boundary, in case the server ever redelivers it.
type responseStreamTransport struct {
	stream        *ResponseStream
	resumedFrom   string
	dedupePending bool
}

func (t *responseStreamTransport) open(ctx context.Context) (*http.Response, error) {
	t.resumedFrom = t.stream.lastEventID
	t.dedupePending = t.resumedFrom != ""
	return t.stream.client.openEventStream(ctx, t.stream.sessionID, t.resumedFrom)
}

func (t *responseStreamTransport) processFrame(f SSEFrame) (Event, bool) {
	if f.ID != "" {
		t.stream.lastEventID = f.ID
	}
	if f.Data == "" {
		return Event{}, false
	}
	e, ok := decodeEvent(f.Data)
	if !ok {
		return Event{}, false
	}
	if t.dedupePending {
		t.dedupePending = false
		if f.ID != "" && f.ID == t.resumedFrom {
			return Event{}, false // duplicate of the resume boundary
		}
	}
	return e, true
}

// Events returns a range-over-func iterator over the run's events. A non-nil error is yielded once,
// last, when the stream fails terminally (a typed APIError on a status error, a ConnectionError on
// an exhausted reconnect). Breaking out of the range closes the transport.
func (s *ResponseStream) Events(ctx context.Context) func(yield func(Event, error) bool) {
	return func(yield func(Event, error) bool) {
		runStream(ctx, &responseStreamTransport{stream: s}, nil, s.maxReconnects, s.backoffBaseMs, s.backoffMaxMs, yield)
	}
}

// FinalResponse drains the stream to its terminal event, then returns the canonical terminal
// Response (a subsequent authenticated retrieve). It shares the single underlying iteration.
func (s *ResponseStream) FinalResponse(ctx context.Context) (*Response, error) {
	var streamErr error
	for _, err := range s.Events(ctx) {
		if err != nil {
			streamErr = err
			break
		}
	}
	if streamErr != nil {
		return nil, streamErr
	}
	return s.client.Responses.Retrieve(ctx, s.responseID)
}

// openEventStream opens the raw SSE response for a session with Bearer auth and, on a resume, the
// Last-Event-ID cursor. It does not retry — ResponseStream owns reconnection.
func (c *Client) openEventStream(ctx context.Context, sessionID, lastEventID string) (*http.Response, error) {
	url := c.baseURL + "/v1/sessions/" + escapePathSegment(sessionID) + "/events"
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, url, nil)
	if err != nil {
		return nil, err
	}
	req.Header.Set("Authorization", "Bearer "+c.apiKey)
	req.Header.Set("API-Version", APIVersion)
	req.Header.Set("Accept", "text/event-stream")
	req.Header.Set("Cache-Control", "no-store")
	if lastEventID != "" {
		req.Header.Set("Last-Event-ID", lastEventID)
	}
	resp, err := c.httpClient.Do(req)
	if err != nil {
		return nil, &ConnectionError{Message: "GET " + url + " failed to reach the server", Cause: err}
	}
	return resp, nil
}
